libyul/optimiser/CommonSwitchCasePrefixMover.cpp

```diff
@@ -0,0 +1,161 @@
+#include <range/v3/algorithm/transform.hpp>
+#include <range/v3/iterator.hpp>
+#include <range/v3/view/drop.hpp>
+#include <range/v3/view/transform.hpp>
+#include <range/v3/view/zip.hpp>
+#include <range/v3/action/transform.hpp>
+
+using namespace std;
+using namespace solidity;
+using namespace solidity::yul;
+
+namespace
+{
+class IdentifierReplacer: public ASTCopier
+{
+public:
+	IdentifierReplacer(map<YulString, YulString> _identifierMap): m_identifierMap(move(_identifierMap)) {}
+	using ASTCopier::operator();
+protected:
+	YulString translateIdentifier(YulString _name) override
+	{
+		if (YulString const* value = util::valueOrNullptr(m_identifierMap, _name))
+			return *value;
+		else
+			return _name;
+	}
+private:
+	map<YulString, YulString> m_identifierMap;
+};
+}
+
+void CommonSwitchCasePrefixMover::run(OptimiserStepContext&, Block& _ast)
+{
+	CommonSwitchCasePrefixMover{}(_ast);
+}
+
+namespace
+{
+
+template<typename Container>
+auto eraseFirst(Container& _container) { return _container.erase(begin(_container)); }
+
+optional<Statement> tryExtractFirstStatement(Identifier const* _identifier, vector<Case>& _cases)
+{
+	Statement& referenceStatement = _cases.front().body.statements.front();
+
+	if (_identifier)
+	{
+		Assignments assignments;
+		visit(assignments, referenceStatement);
+		if (assignments.names().count(_identifier->name))
+			return nullopt;
+	}
+
+	if (!ranges::all_of(
+		_cases | ranges::cpp20::views::drop(1),
+		[&](Case& _case) {
+			return !_case.body.statements.empty() &&
+				SyntacticallyEqual{}(referenceStatement, _case.body.statements.front());
+		}
+	))
+		return nullopt;
+
+	if (VariableDeclaration const* referenceVarDecl = get_if<VariableDeclaration>(&referenceStatement))
+	{
+		for (Case& switchCase: _cases | ranges::cpp20::views::drop(1))
+		{
+			VariableDeclaration* varDecl = std::get_if<VariableDeclaration>(&switchCase.body.statements.front());
+			yulAssert(varDecl, "");
+			yulAssert(varDecl->variables.size() == referenceVarDecl->variables.size(), "");
+
+			static constexpr auto nameFromTypedName = [](TypedName const& _name) { return _name.name; };
+			auto range = ranges::views::zip(
+				varDecl->variables | ranges::views::transform(nameFromTypedName),
+				referenceVarDecl->variables | ranges::views::transform(nameFromTypedName)
+			);
+			std::map<YulString, YulString> identifierReplacement{ranges::begin(range), ranges::end(range)};
+			IdentifierReplacer replacer{{ranges::begin(range), ranges::end(range)}};
+			vector<Statement> newBody;
+			for (auto const& stmt: switchCase.body.statements | ranges::cpp20::views::drop(1))
+				newBody.emplace_back(replacer.translate(stmt));
+			switchCase.body.statements = move(newBody);
+		}
+		Statement result = move(referenceStatement);
+		eraseFirst(_cases.front().body.statements);
+		return result;
+	}
+	else
+	{
+		Statement result = move(referenceStatement);
+		for (Case& switchCase: _cases)
+			eraseFirst(switchCase.body.statements);
+		return result;
```

Comment:
if reference statement is not vardecl then erase first stmt of all switch case blocks.